lib/macs/cbc_block_cipher_mac.dart

```diff
@@ -0,0 +1,218 @@
+     * authentication code half the length of the block size of the cipher.
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     */
+  CBCBlockCipherMac.fromCipher(BlockCipher cipher)
+      : this(cipher, (cipher.blockSize * 8) ~/ 2, null);
+
+  /**
+     * create a standard MAC based on a CBC block cipher. This will produce an
+     * authentication code half the length of the block size of the cipher.
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     * @param padding the padding to be used to complete the last block.
+     */
+  CBCBlockCipherMac.fromCipherAndPadding(BlockCipher cipher, Padding padding)
+      : this(cipher, (cipher.blockSize * 8) ~/ 2, padding);
+
+  /**
+     * create a standard MAC based on a block cipher with the size of the
+     * MAC been given in bits. This class uses CBC mode as the basis for the
+     * MAC generation.
+     * <p>
+     * Note: the size of the MAC must be at least 24 bits (FIPS Publication 81),
+     * or 16 bits if being used as a data authenticator (FIPS Publication 113),
+     * and in general should be less than the size of the block cipher as it
+     * reduces the chance of an exhaustive attack (see Handbook of Applied
+     * Cryptography).
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     * @param macSizeInBits the size of the MAC in bits, must be a multiple of 8.
+     */
+  CBCBlockCipherMac.fromCipherAndMacSize(BlockCipher cipher, int macSizeInBits)
+      : this(cipher, macSizeInBits, null);
+
+  /**
+     * create a standard MAC based on a block cipher with the size of the
+     * MAC been given in bits. This class uses CBC mode as the basis for the
+     * MAC generation.
+     * <p>
+     * Note: the size of the MAC must be at least 24 bits (FIPS Publication 81),
+     * or 16 bits if being used as a data authenticator (FIPS Publication 113),
+     * and in general should be less than the size of the block cipher as it
+     * reduces the chance of an exhaustive attack (see Handbook of Applied
+     * Cryptography).
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     * @param macSizeInBits the size of the MAC in bits, must be a multiple of 8.
+     * @param padding the padding to be used to complete the last block.
+     */
+  CBCBlockCipherMac(BlockCipher cipher, int macSizeInBits, Padding padding)
+      : _cipher = new CBCBlockCipher(cipher),
+        _macSize = macSizeInBits ~/ 8,
+        _padding = padding {
+    if ((macSizeInBits % 8) != 0) {
+      throw new ArgumentError("MAC size must be multiple of 8");
+    }
+
+    _mac = new Uint8List(cipher.blockSize);
+
+    _buf = new Uint8List(cipher.blockSize);
+    _bufOff = 0;
+  }
+
+  @override
+  String get algorithmName {
+    String paddingName = _padding != null ? "/${_padding.algorithmName}" : "";
+    return "${_cipher.algorithmName}_CMAC${paddingName}";
```

Comment:
With an underscore?

Reply:
I think this is the nit, and you’d expect a slash? I think there would be a naming conflict with the other CMAC algorithmName. I’m not sure why there are two slightly different CMAC implementations in BouncyCastle, or why they don’t pass the same test vectors (with default parameters, at least)

If you think it is appropriate to include both, how would you like to see the naming?

Reply:
Meh, I'm not sure here anyway. I'll just take your word for it :)